settings/src/components/backup-codes.js

```diff
@@ -85,20 +91,33 @@ function Setup( { setRegenerating } ) {
 
 			<p>Please print the codes and keep them in a safe place.</p>
 
-			<CodeList codes={ backupCodes } />
-
-			<Notice status="warning" isDismissible={ false }>
-				<Icon icon={ warning } className="wporg-2fa__print-codes-warning" />
-				Without access to the one-time password app or a backup code, you will lose access
-				to your account. Once you navigate away from this page, you will not be able to view
-				these codes again.
-			</Notice>
-
-			<CheckboxControl
-				label="I have printed or saved these codes"
-				checked={ hasPrinted }
-				onChange={ setHasPrinted }
-			/>
+			{ error ? (
+				<>
+					{ error.code === 'revalidation_required' && <RevalidateModal /> }
```

Comment:
Agreed. I would lean towards using an `apiFetch` middleware: https://www.npmjs.com/package/@wordpress/api-fetch.

Reply:
Tried using `apiFetch` middleware, but it didn't go well. `Context` couldn't be accessed in the middleware and the RevalidateModal needs it. I've tried several other ways and found out that one seems to work with a quick test. I'll open another PR for it after finishing the rest part of it. And for this PR, I'll `remove the revalidation stuff here, and just have the Notice to catch any kind of API error`. Thanks for bringing this up, it's a really good point. And thanks Steve for sharing a potential solution! 👍

Reply:
So I understand it correctly, you wanted to set the error within the app context and you couldn't?

Reply:
[This](https://github.com/WordPress/wporg-two-factor/compare/trunk...try/revalidation-handler) is the branch I tried with `apiFetch` middleware, it's in the status that can be tested directly. I was trying to add a specific handler for revalidation using middleware, making it render the [components](https://github.com/WordPress/wporg-two-factor/compare/trunk...try/revalidation-handler#diff-ff88c76f4d9fc57f36f88dc481260c2d5ecd496149970a72162c81660f39fa2cR30-R37) we want if `'revalidation_required' !== error.code`. However, as `createRoot` or any other way that dynamically renders components would build a new tree that's unable to access the `Context` data, it would lead to an error message. This would work if only the component [Notice](https://github.com/WordPress/wporg-two-factor/compare/trunk...try/revalidation-handler#diff-ff88c76f4d9fc57f36f88dc481260c2d5ecd496149970a72162c81660f39fa2cR33-R36) is rendered. 

I also tried setting variables or flagging first when the error occurs, and then have components detect these to trigger the re-rendering accordingly. However, this PR might be more [straightforward](https://github.com/WordPress/wporg-two-factor/pull/199).